src/ITensorChainRules/ITensorChainRules.jl

```diff
@@ -3,25 +3,185 @@
 using ChainRulesCore
 using ITensors
 
+# Needed for defining the rule for `adjoint(A::ITensor)`
+# which currently doesn't work by overloading `ChainRulesCore.rrule`
+using ZygoteRules: @adjoint
+
 function ChainRulesCore.rrule(::typeof(getindex), x::ITensor, I...)
   y = getindex(x, I...)
   function getindex_pullback(ȳ)
-    x̄ = ITensor(inds(x))
+    # TODO: add definition `ITensor(::Tuple{}) = ITensor()`
+    # to ITensors.jl so no splatting is needed here.
+    x̄ = ITensor(inds(x)...)
     x̄[I...] = ȳ
     Ī = broadcast(_ -> NoTangent(), I)
     return (NoTangent(), x̄, Ī...)
   end
   return y, getindex_pullback
 end
 
-function ChainRulesCore.rrule(::typeof(*), x1::ITensor, x2::ITensor)
+# Specialized version in order to avoid call to `setindex!`
+# within the pullback, should be better for taking higher order
+# derivatives in Zygote.
+function ChainRulesCore.rrule(::typeof(getindex), x::ITensor)
+  y = x[]
+  function getindex_pullback(ȳ)
+    x̄ = ITensor(ȳ)
+    return (NoTangent(), x̄)
+  end
+  return y, getindex_pullback
+end
+
+function setinds_pullback(ȳ, x, a...)
+  x̄ = ITensors.setinds(ȳ, inds(x))
+  ā = broadcast(_ -> NoTangent(), a)
+  return (NoTangent(), x̄, ā...)
+end
+
+for fname in (
+  :prime,
+  :setprime,
+  :noprime,
+  :replaceprime,
+  :swapprime,
+  :addtags,
+  :removetags,
+  :replacetags,
+  :settags,
+  :swaptags,
+  :replaceind,
+  :replaceinds,
+  :swapind,
+  :swapinds,
+)
+  @eval begin
+    function ChainRulesCore.rrule(::typeof($fname), x::ITensor, a...)
+      y = $fname(x, a...)
+      function f_pullback(ȳ)
+        return setinds_pullback(ȳ, x, a...)
+      end
+      return y, f_pullback
+    end
+  end
+end
+
+# TODO: This is not being called by Zygote for some reason,
+# using a Zygote overload directly instead. Figure out
+# why, maybe raise an issue.
+#function ChainRulesCore.rrule(::typeof(adjoint), x::ITensor)
+#  y = prime(x)
+#  function adjoint_pullback(ȳ)
+#    return setinds_pullback(ȳ, x)
+#  end
+#  return y, adjoint_pullback
+#end
+
+@adjoint function Base.adjoint(x::ITensor)
+  y = prime(x)
+  function setinds_pullback(ȳ)
+    x̄ = ITensors.setinds(ȳ, inds(x))
+    return (x̄,)
+  end
+  return y, setinds_pullback
+end
+
+function _contract_pullback(ȳ, x1, x2)
```

Comment:
What's the reason for factoring this function out from _rrule? I see this function is only called once so confused.

Reply:
Probably no need, I think just a leftover of how I had organized the code previously before then factoring out a common `_rrule`.

Actually in general I wished that other packages would factorize out pullback definitions more often (for example in linear algebra functions like `svd`, where I may want to use the `svd` pullback definition within my own ITensor pullback function) but this may not be a case where that is relevant.